[#14] 주문

seohoon-dev opened the pull request:
- 주문 등록
- 주문 시 결제 진행
- 주문 완료 후 장바구니 비움

seohoon-dev committed:
[#14] 주문

- 주문 등록
- 주문 시 결제 진행
- 주문 완료 후 장바구니 비움

seohoon-dev commented on src/main/resources/mapper/OrderMapper.xml:
```diff
@@ -0,0 +1,77 @@
+            <result property="price" column="price"/>
+            <result property="amount" column="amount"/>
+        </collection>
+    </resultMap>
```
regDate 값을 불러오는 과정에서 timestamp 오류가 나는데 원인을 잘 모르겠습니다..
resultMap 내부에 collection 부분을 주석처리하고 호출하면 OrderHistory 부분은 정상적으로 출력하는데 collection만 추가하면 아래 오류가 발생하네요..
mysql과 java 데이터타입은 timestamp로 동일합니다.

> Cause: java.sql.SQLDataException: Unsupported conversion from TIMESTAMP to java.lang.Integer

f-lab-dev replied:
일반적으로 TimeStamp 자료형은 잘 쓰지 않습니다. `java.util.Date`를 사용해보시는건 어떨까요?

seohoon-dev commented on src/main/java/com/hoon/foodrocket/config/PaymentConfig.java:
```diff
@@ -0,0 +1,22 @@
+        map.put(PaymentMethod.CARD, new CardPaymentProcess());
+        map.put(PaymentMethod.PHONE, new PhonePaymentProcess());
+        map.put(PaymentMethod.KAKAO, new KakaoPaymentProcess());
+
```
결제 모듈은 로그인 검증과 동일하게 Map을 사용했습니다.
저번에 말씀하셨던 결제 모듈이 추가될 경우 Config에 등록하지 않을 수 있다는 문제를 고민해봤지만 임시방편으로 주석을 달아주는 것 밖에 생각 나질 않네요..

f-lab-dev commented on src/main/java/com/hoon/foodrocket/application/CartService.java:
```diff
@@ -36,4 +36,8 @@ public void deleteItem(Long id, String loginUserEmail) {
     }
+
+    public void clearItem(String loginUserEmail) {
+        redisTemplate.delete(loginUserEmail);
```
이거로 미루어보아 레디스 키로 이메일을 그대로 쓰신 것 같습니다. 이런 경우엔 `cart:{email}` 이런식으로 키를 잡아주는게 구분하기 좋습니다~

f-lab-dev commented on src/main/java/com/hoon/foodrocket/application/OrderService.java:
```diff
@@ -0,0 +1,62 @@
+     * <p>
+     * 2. 디비에 주문 등록
+     * 오더 테이블에 주문 정보를 등록한다.
+     * 오더 메뉴 테이블에 오더의 메뉴 정보를 등록한다.
```
여기서는 주문 등록을 1번으로 해주는게 나을 수도 있습니다~
결제가 성공하고 주문에 실패하면 결제를 취소시켜야하는데 그거는 DB롤백만으로 해결이 안됍니다. 그런 경우는 rollback 이벤트를 받아서 결제취소 로직을 돌려야하는데 그렇게 처리하시거나 혹은 순서를 바꿔주는게 좋을 것 같네요~

f-lab-dev commented on src/main/java/com/hoon/foodrocket/application/OrderService.java:
```diff
@@ -0,0 +1,62 @@
+     * @param order
+     * @param loginUserEmail
+     */
+    @Transactional
```
이거는 참고사항이지만 주문과 결제는 독립적으로 일어날 수 있습니다.
주문을 먼저 한 다음, 결제완료 여부만 표시해주면 되니까요

f-lab-dev commented on src/main/java/com/hoon/foodrocket/application/payment/PaymentService.java:
```diff
@@ -0,0 +1,24 @@
+    }
+
+    public boolean process(Order order) {
+        // 결제 모듈이 추가될 경우 PaymentConfig.java 에 추가해야함
```
일단 제가 말씀드린것과 비슷하게 구현을 잘 하셨네요. 이제 전략패턴을 검색해보시고 바꿔서 적용해보셔도 좋을 것 같네요~

f-lab-dev commented on src/main/java/com/hoon/foodrocket/domain/Order.java:
```diff
@@ -0,0 +1,43 @@
+import javax.validation.constraints.NotNull;
+import java.util.List;
+
+@Builder
```
이 클래스에 빌더 패턴을 사용하신 이유가 있을까요?

f-lab-dev commented on src/main/resources/mapper/OrderMapper.xml:
```diff
@@ -0,0 +1,77 @@
+    <select id="getOrderHistoryList" resultMap="orderHistoryMap">
+        select a.id ordersId, a.restaurant_name restaurantName, a.payment_amount paymentAmount, a.reg_date regDate,
+        b.id orderMenusId, b.name, b.price, b.amount
+        FROM orders a LEFT OUTER JOIN order_menus b
```
여기서는 INNER JOIN이 아니라 OUTER JOIN을 사용한 이유가 있으실까요? 

쿼리마다 인덱스를 타는지 여부, 조인 종류 선택 등 쿼리 설계에 대한 의도를 주석으로 남겨주시면 좋을 것 같습니다~

seohoon-dev committed:
리뷰 내용 반영

- 패키지명 변경(application -> service)
- 주문, 결제 순서 변경
- 결제 파트에 전략패턴 적용
- 주문 내역 시 커서 기반 페이징 적용
- 주문 상세 내역 추가

f-lab-dev commented on src/main/java/com/hoon/foodrocket/controller/OrderController.java:
```diff
@@ -2,9 +2,11 @@
-import com.hoon.foodrocket.application.OrderService;
-import com.hoon.foodrocket.domain.Order;
-import com.hoon.foodrocket.domain.OrderHistory;
+import com.hoon.foodrocket.domain.order.*;
```
보통 컨벤션에서 `*`을 이용한 import는 지양하는 편입니다. 코딩 컨벤션 툴을 사용해보시면 어떨까요?

f-lab-dev commented on src/main/java/com/hoon/foodrocket/controller/OrderController.java:
```diff
@@ -20,21 +22,62 @@
     @LoginType(level = UserAuthorityLevel.USER)
-    @GetMapping
-    public List<OrderHistory> list(HttpSession session) {
+    @GetMapping("/list")
```
보통 이런 `/list`는 붙여주지 않습니다. `/orders` 라는 URL안에 이미 list라는 뜻이 있기 때문이죠.
혹시 붙여주신 이유가 있을까요? 이걸 꼭 붙여야했을 상황이라면 URL설계가 뭔가 잘못되었을 가능성이 있습니다.

seohoon-dev replied:
제가 url을 잘못 붙였습니다. 확인해보니 프로젝트 내부에 저런식으로 /list를 붙여준 적이 없더군요. 주의하겠습니다.

f-lab-dev commented on src/main/java/com/hoon/foodrocket/controller/OrderController.java:
```diff
@@ -20,21 +22,62 @@
     @LoginType(level = UserAuthorityLevel.USER)
-    @PostMapping
-    public HttpStatus create(@RequestBody Order order, HttpSession session) {
+    @GetMapping("/detail")
```
이런 경우는 `/orders/{id}/detail` 이렇게 가는게 좋습니다

seohoon-dev replied:
이부분도 다른 곳은 `/{id}`로 처리했는데 이날 정신이 없었나봅니다..

f-lab-dev commented on src/main/java/com/hoon/foodrocket/controller/OrderController.java:
```diff
@@ -20,21 +22,62 @@
+    }
+
+    @LoginType(level = UserAuthorityLevel.USER)
+    @PostMapping("/card")
```
여기는 `/orders/payments/card` 이렇게 가는게 어떨까요?
전체적인 URL 설계쪽을 다시 보면 좋을 것 같습니다. 
https://docs.microsoft.com/ko-kr/azure/architecture/best-practices/api-design

f-lab-dev reviewed (commented):
PR이 너무 커져서 이전에 말씀드린것처럼 이슈를 세밀하게 나누는게 필요합니다.
주문은 너무 큰 덩어리(에픽)이고, 주문에도 작은 로직(이슈)들이 많이 모여있습니다.

예를 들면 주문에도 결제, 주문요청, 주문조회 등이 있겠죠. 이들을 세밀하게 나눠주시는게 코드관리에 좋습니다.

github 이슈에 다 세밀하게 나눠서 올려주세요~

seohoon-dev committed:
Merge branch 'issue/15' of https://github.com/f-lab-edu/food-rocket into issue/14

seohoon-dev committed:
리뷰 내용 반영

- single import 방식으로 변경
- url 변경

f-lab-dev reviewed (approved):
쿼리에 join 등 쿼리의 의도에 대한 주석만 작성해주시면 될 것 같네요

seohoon-dev committed:
리뷰 내용 반영

- 쿼리에 주석 작성
